Make portable implementation `const`

nazar-pc opened the pull request:
I was going to use BLAKE3 in `const` environment, but it doesn't support such use case yet.

This relatively small PR refactors portable implementation to make it work in `const` environment.

There is a bit of trivially verifiable `unsafe` due to nicer APIs not yet available in `const` environment, but these all work on stable.

`arrayref` were hard to read and used slicing that is not compatible with `const`, so I replaced it with a couple simple macros where it was necessary and left other usages as is.

---

UPD: I decided to implement the whole thing, added `const_hash`, `const_keyed_hash` and `const_derive_key` functions by copying the originals and modifying them to use `const fn` portable implementation.

I didn't add `ConstHasher`, though it is possible too if someone needs it.

Resolves https://github.com/BLAKE3-team/BLAKE3/issues/440

nazar-pc committed:
Add `const fn` functions: `const_hash`, `const_keyed_hash` and `const_derive_key`

devinrsmith commented on src/platform.rs:
```diff
@@ -473,74 +472,96 @@ pub fn sse2_detected() -> bool {
+}
+
 #[inline(always)]
 pub fn words_from_le_bytes_32(bytes: &[u8; 32]) -> [u32; 8] {
```
Missing `const` here?

nazar-pc replied:
Wasn't needed, but will add for consistency with other functions

devinrsmith commented on src/portable.rs:
```diff
@@ -120,7 +119,7 @@ pub fn compress_xof(
 }
 
-pub fn hash1<const N: usize>(
+const fn hash1<const N: usize>(
```
Removal of `pub` intended?

nazar-pc replied:
Not really, but since there are a few other backends with the same function and in all cases it is internal implementation without `pub` (here too), I decided to remove it for consistency too. Happy to restore if maintainers have strong opinion on this one.

devinrsmith reviewed (commented):
FYI: I'm not a maintainer of this library, but I do like to keep up to date with some of the going-ons.

nazar-pc reviewed (commented):
Thanks for review, appreciate it!

nazar-pc commented:
@oconnor663 it'd be great to get your attention here, especially since the changes are fairly simple

nazar-pc commented:
Ping :slightly_smiling_face:

nazar-pc commented:
I have implemented public const functions too and added them to tests against reference implementation, everything passes. Just rebased against `master` to pull Rust 1.78.0 to get some of the features that were previously unstable on older version of Rust.

@oconnor663 would be great if you can find time to look at this, especially since all tests are passing.

nazar-pc commented:
Hm... I see https://github.com/BLAKE3-team/BLAKE3/commit/a31e519869d5751370f50c39a99340660ee95bf7 says that true MSRV is 1.66.0, if that is the case I can put new features behind feature flag or maybe MSRV can be increased (1.77.0 is what is needed here I believe, which was released almost a year ago).

oconnor663 commented:
Apologies for not reviewing this more promptly. You might already have noticed that I just pushed the CI-tested MSRV to 1.78 (for the `ignore` crate).

nazar-pc commented:
Just discovered 1.83.0 is what is needed for `ptr::copy_to_nonoverlapping()`

nazar-pc commented:
Okay, so `&mut` is not allowed in `const fn` until current stable 1.85.0. Sorry for the noise here, I added `const` feature that enabled these new APIs, retaining support for older versions of Rust.

oconnor663 commented:
I'm leaning against taking these changes. It looks like a lot of complicated logic has to be duplicated, and much of that logic (like `compress_subtree_wide`) is only as complicated as it is because it supports SIMD parallelism, which the `const` version gets no benefit from. I wonder if it could make more sense to fork the `reference_impl` rather than the highly optimized implementation?

That said, my gut level, haven't-spent-much-time-with-the-feature impression of const generics is that it's still too early to rely heavily on them. Errors like this feel silly to me:

```
error[E0015]: cannot use `for` loop on `std::ops::Range<usize>` in constant functions
  --> reference_impl.rs:68:14
   |
68 |     for i in 0..16 {
   |              ^^^^^
```

Another minimim-viable-feature for me would be basic arithmetic in const parameters, which doesn't work today:

```rust
struct Hasher<const LEN: usize> {
    state: [u8; LEN * 2],  // error: generic parameters may not be used in const operations
}
```

Does all that sound reasonable?

nazar-pc committed:
Rewrite `platform` functions to `const fn`

nazar-pc committed:
Refactor `portable` to be `const fn`-friendly (but not const yet)

oconnor663 commented:
I'm still mostly against making this change within this crate, but I think a `const` fork of the portable implementation or the reference implementation makes sense, and I'd be happy to review it and link to it from our README. I think there are enough `const` functions already in place on `blake3::Hash` that the return types could interoperate, but if we need to make tweaks there that could also make sense.
